ulauncher/modes/extensions/extension_environment.py

```diff
@@ -0,0 +1,58 @@
+"""
+!EXPERIMENTAL: this module intends to provide an implementation of automatic
+!installation and virtual environment management/setup automation of ulauncher
+!extensions installed using one way or another.
+"""
+
+from pathlib import Path
+from os.path import exists
+
+
+class ExtensionEnvironment:
+    def __init__(self, ext_dir: Path, venv_map: dict[str, set]) -> None:
+        self.ext_dir: Path = ext_dir
+        self.venv_map: dict[str, set[str]] = {
+            "test": set(
+                    [
+                        "hello==1.0"
+                    ]
+                )
+        }
+
+    def check_requirements(self) -> bool:
+        """check_requirements
+
+        Args:
+            ext_dir (Path): Path of the extension dir downloaded/cloned.
+
+        Returns:
+            bool: whether a requirements.txt file exists for the extension.
+        """
+
+        if exists(self.ext_dir / "requirements.txt"):
+            return True # it should be requirements.txt only, not requirement.txt
+        return False
+
+    def prep_extension(self):
+        req_path: Path = self.ext_dir / "requirements.txt"
+
+        if not self.check_requirements():
+            self.create_sole_venv()
+
+        with open(req_path, "r", encoding="utf-8") as req_file:
+            dependencies: set[str] = set(req_file.readlines())
```

Comment:
We should definitely not attempt to parse dependencies ourselves, this should be offloaded entirely to `pip`

Reply:
the reasoning behind that is that it will be stored into a json.txt, and it will be later read when another extension is needed to be installed. if the requirements.txt of the new extension is a subset or is similar to the another one, then we can just install that similar extension to the same venv